services/ns_net/allocation_builder.cc

```diff
@@ -0,0 +1,77 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/services/ns_net/allocation_builder.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+namespace mojo {
+
+AllocationBuilder::AllocationBuilder()
+    : buffer_(nullptr), buffer_length_(0), data_length_(0) {}
+
+AllocationBuilder::~AllocationBuilder() {
+  free(buffer_);
+}
+
+uint32_t AllocationBuilder::Size() const {
+  return data_length_;
+}
+
+static inline uint32_t RoundPowerOfTwo(uint32_t x) {
+  if (x == 0) {
+    return 1;
+  }
+
+  --x;
+
+  x |= x >> 1;
+  x |= x >> 2;
+  x |= x >> 4;
+  x |= x >> 8;
+  x |= x >> 16;
+
+  return x + 1;
+}
+
+bool AllocationBuilder::Append(const uint8_t* data, uint32_t length) {
```

Comment:
I don't think `std::vector` is ideal for dealing with allocations of data for the following reasons. Some are practical, others more aesthetic:
* I don't see a way of handling allocation failures during resizes of the container (at least not without enabling exceptions and handling `std::bad_alloc`). This implementation correctly refuses to resize the container if `realloc` fails, and, returns an error to the caller (decorated with `FTL_WARN_UNUSED_RESULT` so it's hard to ignore). We use this indication later in the `ns_net` implementation to send an error response with a descriptive message that says we ran out of memory. This allocation builder can be used to service potentially massive response bodies (more on this later). I have wondered if there is some indication as to whether there was an allocation failure from an `std::vector` `resize` by checking if the `data()` pointer after is dereferenceable but I haven't found a definitive answer (there very well might be a way to do this, i'd rather have these rules be clear). I don't know if it just asserts or corrupts data.
* Both the `AllocationBuilder` and the `Allocation` are not copy constructible or assignable. When dealing with `std::vector` I am always weary of missing a `std::move` and incurring extra overhead do to a copy. Its just harder to make a mistake here.
* To move data into the container, doing the `resize`, accessing the base pointer via `data()` or `front()` and then `memcpy`ing it in seems ... icky. Won't the resize cause an unnecessary zero'ing out of the vector elements? Or do we move the allocation one item at a time.
* I wanted a way to steal the allocation once it was built and pass it over to the data pipe flusher. Using `std::vector, `I couldn't find a way to do this without incurring another copy. This one is admittedly not as big of an issue because I could have just made that take a vector and moved it in.

About the massive response bodies: Even though we now handle allocation failures and tell the caller of this condition, this is far from ideal. We probably want to gate the rate at which we read from the underlying socket rather that read as much as we can in the `ns_net` layer and then try to be as efficient as possible in transferring over that data to the caller. I was trying to do the latter with the ring buffer but now I think I will drop that idea. This issue makes me nervous about this interface when it comes to dealing with extremely large response bodies.